spec/brainstem/presenter_collection_spec.rb

```diff
@@ -159,6 +159,73 @@ class MyException < Exception; end
           expect(result['count']).to eq(Workspace.count)
         end
       end
+
+      describe "pagination keys" do
+        let(:max_per_page) { nil }
+        let(:result) { @presenter_collection.presenting("workspaces", params: params, max_per_page: max_per_page) { Workspace.unscoped } }
+
+        describe "page_number" do
+          describe "when page is provided" do
+            let(:params) { { page: 2 } }
+
+            it "indicates the provided page" do
+              expect(result["page_number"]).to eq(2)
+            end
+          end
+
+          describe "when page is not profided" do
+            let(:params) { {} }
+
+            it "indicates the first page" do
+              expect(result["page_number"]).to eq(1)
+            end
+          end
+        end
+
+        describe "page_count" do
+          describe "when per_page is provided" do
+            let(:params) { { per_page: 4 } }
+
+            it "calculates the correct page count" do
+              expect(result["page_count"]).to eq(2)
+            end
+          end
+
+          describe "when per_page is not provided" do
+            let(:params) { {} }
+
+            it "calculates the correct page count based on the default page size" do
+              expect(result["page_count"]).to eq(3)
+            end
+          end
+        end
+
+        describe "page_size" do
+          [
+            # per_page  default_per_page  max_per_page  default_max_per_page  expected   message
+            [       10,               20,          100,                  200,       10,  "per page < max"],
+            [      nil,               20,          100,                  200,       20,  "no per page and default < max"],
+            [      nil,              200,          100,                  200,      100,  "no per page and default > max"],
+            [      150,               20,          100,                  200,      100,  "per page > max"],
+            [      150,               20,          nil,                  200,      150,  "no max and per page < default max"],
+            [      250,               20,          nil,                  200,      200,  "no max and per page > default max"],
+          ].each do |per_page, default_per_page, max_per_page, default_max_per_page, expected, message|
+            describe "for foo" do
+              let(:params) { { per_page: per_page } }
+              let(:max_per_page) { max_per_page }
+
+              before do
+                @presenter_collection.default_per_page = default_per_page
+                @presenter_collection.default_max_per_page = default_max_per_page
+              end
+
+              it "handles '#{message}'" do
+                expect(result["page_size"]).to eq(expected)
+              end
```

Comment:
@shirish-pampoorickal @bennapp @juanca , I'm not loving this test...

But I **_really_** don't want to write out all the different cases manually. Any thoughts or other ideas?

Reply:
I think this is fine. 

I guess a more important question(s) to ask: 

- Why do we need to test all these cases? 
- Why can we not guarantee a well behaved pagination object?

Reply:
I was thinking we'd want to test all the logic of the calculation. Also, not sure what you mean be guaranteeing a well-behaved pagination object.

Reply:
If the presenter was using a Pagination module/object and that *thing* always enforced sane pagination values then we wouldn't need to deal with all these test cases.

Reply:
Right, that makes sense. We would still want to test all the cases somewhere, though, just not in here.

I think we're on the same page, though

Reply:
I think we can drop the `nil` cases by using defaults and I think we can drop the invalid states by using validations. 

The tests can be more streamlined and get "to the point" by only asserting simple transformations (e.g. given a `per_page` bigger than the max will always set it to the max) versus these integration tests. 

That said: I may not have all the context on all these states and tests. All of these comments are more conversation starters. Would appreciate other contributors chiming in.

Reply:
We need the `nil` cases because there's a bunch of default handling going on.

For example, if no `per_page` is provided, then the `default_per_page` is used. And if no `max_per_page` is provided, there's a `default_max_per_page` that gets used.

The nils end up being important to the logic. Now, if the logic gets moved out of here they don't need to say in here, but they need to be tested somewhere.